### 01 Importing Libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02 Import Data

In [4]:
path = r'C:\Users\isava\OneDrive\Documents\CareerFoundry\Data Immersion\PythonFundamentals\Instacart Basket Analysis'
path

'C:\\Users\\isava\\OneDrive\\Documents\\CareerFoundry\\Data Immersion\\PythonFundamentals\\Instacart Basket Analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','orders_wrangled.csv'), index_col = False)
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','products_checked.csv'), index_col = False)
df_ords_prior = pd.read_csv(os.path.join(path,'02 Data', 'Original Data','orders_products_prior.csv'), index_col = False)

### Save Memory

In [7]:
df_ords.drop(columns = {'Unnamed: 0'}, inplace = True)

In [8]:
df_ords.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_users_prior_order'],
      dtype='object')

In [9]:
df_ords.rename(columns = {'order_dow':'orders_day_of_week', 'order_hour_of_day':'order_time_of_day'}, inplace = True)

In [10]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   order_id                      int64  
 1   user_id                       int64  
 2   order_number                  int64  
 3   orders_day_of_week            int64  
 4   order_time_of_day             int64  
 5   days_since_users_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


In [11]:
#Change types for ords data set

df_ords['order_id']= df_ords['order_id'].astype('int32')
df_ords['user_id'] = df_ords['user_id'].astype('int32')
df_ords['order_number']=df_ords['order_number'].astype('int8')
df_ords['orders_day_of_week']=df_ords['orders_day_of_week'].astype('int8')
df_ords['order_time_of_day']=df_ords['order_time_of_day'].astype('int8')
df_ords['days_since_users_prior_order']=df_ords['days_since_users_prior_order'].astype('float16')

In [12]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   order_id                      int32  
 1   user_id                       int32  
 2   order_number                  int8   
 3   orders_day_of_week            int8   
 4   order_time_of_day             int8   
 5   days_since_users_prior_order  float16
dtypes: float16(1), int32(2), int8(3)
memory usage: 42.4 MB


In [13]:
df_ords_prior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB


In [14]:
# Change types for ords prods data set 

df_ords_prior['product_id'] =df_ords_prior['product_id'].astype('int32')
df_ords_prior['reordered']=df_ords_prior['reordered'].astype('int8')
df_ords_prior['add_to_cart_order']=df_ords_prior['add_to_cart_order'].astype('int32')
df_ords_prior['order_id']=df_ords_prior['order_id'].astype('int32')

In [15]:
df_ords_prior.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int32
 1   product_id         int32
 2   add_to_cart_order  int32
 3   reordered          int8 
dtypes: int32(3), int8(1)
memory usage: 402.1 MB


In [16]:
df_prods.columns

Index(['Unnamed: 0', 'product_id', 'product_name', 'aisle_id', 'department_id',
       'prices'],
      dtype='object')

In [17]:
# Further columns from other data sets
df_prods.drop(columns = {'Unnamed: 0'}, inplace = True)

In [18]:
df_prods['aisle_id'] = df_prods['aisle_id'].astype('int8')
df_prods['department_id'] = df_prods['department_id'].astype('int8')
#df_prods['prices'] = df_prods['prices'].astype('float16')

In [19]:
df_prods.describe()

,product_id,aisle_id,department_id,prices
count,49672.000000,49672.000000,49672.000000,49672.000000
mean,24850.349775,58.320664,11.728942,9.993282
std,14340.705287,51.109052,5.850779,453.615536
min,1.000000,-128.000000,1.000000,1.000000
25%,12432.750000,28.000000,7.000000,4.100000
50%,24850.500000,65.000000,13.000000,7.100000
75%,37268.250000,98.000000,17.000000,11.100000
max,49688.000000,127.000000,21.000000,99999.000000


# 03 Merging Data

In [21]:
df_ords_prior.head()

,order_id,product_id,add_to_cart_order,reordered
0,2,33120,1,1
1,2,28985,2,1
2,2,9327,3,0
3,2,45918,4,1
4,2,30035,5,0


In [22]:
df_ords_prior.shape

(32434489, 4)

We want to merge df_ords prior with df_ords based on the order_id. 

In [24]:
df_ords.head()

,order_id,user_id,order_number,orders_day_of_week,order_time_of_day,days_since_users_prior_order
0,2539329,1,1,2,8,NaN
1,2398795,1,2,3,7,15.0
2,473747,1,3,3,12,21.0
3,2254736,1,4,4,7,29.0
4,431534,1,5,4,15,28.0


In [25]:
df_merged_large = df_ords.merge(df_ords_prior, on = 'order_id', indicator = True)

In [26]:
df_merged_large['_merge'].value_counts()

_merge
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

The resulting dataframe has 32,434,489 rows.

# 04 Export Data Frame

In [29]:
#df_merged_large.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.csv'))
df_merged_large.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_combined.pkl'))